mono/metadata/threads.c

```diff
@@ -2699,16 +2704,23 @@ mono_thread_resume (MonoInternalThread *thread)
 }
 
 void
-ves_icall_System_Threading_Thread_Resume (MonoThread *thread)
+ves_icall_System_Threading_Thread_Resume (MonoThreadObjectHandle thread_handle, MonoError *error)
 {
-	if (!thread->internal_thread) {
-		mono_set_pending_exception (mono_get_exception_thread_state ("Thread has not been started, or is dead."));
+	// Internal threads are pinned so shallow coop/handle.
+	MonoInternalThread * const internal_thread = thread_handle_to_internal_ptr (thread_handle);
+	gboolean exception = FALSE;
+
+	if (!internal_thread) {
+		exception = TRUE;
 	} else {
-		LOCK_THREAD (thread->internal_thread);
-		if (!mono_thread_resume (thread->internal_thread))
-			mono_set_pending_exception (mono_get_exception_thread_state ("Thread has not been started, or is dead."));
-		UNLOCK_THREAD (thread->internal_thread);
+		LOCK_THREAD (internal_thread);
+		if (!mono_thread_resume (internal_thread))
```

Comment:
Move exception out of the lock. Similar cases are not locked, and the raise is not on the same thread (?) as the locking -- or can it be on self?

Reply:
Could add an assert to find out.

Reply:
Resetting CI but ok.

Reply:
Or just put it back in the lock?

Reply:
Raise under lock or not? It is always best to hold locks as short of a time/range as possible, and no shorter. The key is finding the minimum sufficient way.

Reply:
That's true, I guess there's a strdup at least.